### ISSUE TITLE
Type error on .include()

### ISSUE THREAD

**Issue by gerardolima** [hapijs/code#162]
#### Support plan

* *is this issue currently blocking your project?* (yes/no): no
* *is this issue affecting a production system?* (yes/no): yes

#### Context

* *node version*: 14.16.0
* *module version with issue*:  8.0.2
* *last module version without issue*: unknown
* *environment* (e.g. node, browser, native): node
* *used with* (e.g. hapi application, another framework, standalone, ...): AWS Lambda
* *any other relevant information*: Using TypeScript version 4.2.3

#### What are you trying to achieve or the steps to reproduce?

```ts
// from https://hapi.dev/module/code/api/?v=8.0.3
expect({ a: 1, b: 2, c: 3 }).to.include({ a: 1 });
```

#### What was the result you got?
```
Argument of type '{ a: number; }' is not assignable to parameter of type 'string | string[] | { a: number; b: number; c: number; } | { a: number; b: number; c: number; }[]'.
  Type '{ a: number; }' is not assignable to type 'string'.ts(2345)
```
![image](https://user-images.githubusercontent.com/4498885/113744203-6841cd80-96fc-11eb-8124-c7ee1a98eb8d.png)

#### What result did you expect?
No type error.

**Comment by kanongil** [comment]
That is a constructed test, that should never occur as is. The expect value would normally be some property.

In this case Code uses the inferred type from the expect object (`{ a: number, b: number, c: number }`) , to validate against the include arg. It would pass the type check, if the type actually matched, eg. `{ a: 0, b: 0, c: 0 }`.

If you don't want to use the inferred type, you can specify it using generics, as in the test here: https://github.com/hapijs/code/blob/1aa2fc79c4e9032c4edc54c0f58322aac5a0a716/test/index.ts#L132

The error message is highly confusing though, and could be improved in Code using overloads.

**Comment by gerardolima** [comment]
hey, @kanongil, thank you for the answer! I wonder if using `Partial<T>` to check for properties in the subject value `<T>` would be a mode _intuitive_ signature, as I try to demonstrate, below:

```ts
// on @hapi/code/lib/index.d.ts (v 8.0.2)
include(values: string | string[] | T | T[]): Assertion<T>;

// suggestion
include(values: string | string[] | Partial<T> | Partial<T>[]): Assertion<T>;
```

What do you think about it?

**Comment by kanongil** [comment]
Thanks for the input. I thought about the same thing, which handles this issue, but was not sure that it actually improves things. I would need some real world use-cases to better judge.

**Comment by gerardolima** [comment]
My case was very alike the sample: asserting the value for `body` property of a HTTP response; something like the following:
```ts
expect(response).to.include({ body: 'hello world' })
```

As I couldn't compile, I had to change code to test for the full response:
```ts
expect(response).to.equal({ body: 'hello world', status: 200 })
```

**Comment by kanongil** [comment]
Thanks, that looks legit. Working on a patch to fix it along with some other issues.

**Comment by Nargonath** [comment]
I was going to suggest `Partial` as well so looks good to me.

**Comment by kanongil** [comment]
I did a fix in #163, where I made the `Partial` recursive, and also allow unknown properties on the base object.
